File: services/email/src/routes/email.ts
import { Router, Request, Response } from 'express';
import { z } from 'zod';
import { emailService } from '../services/EmailService';
import { VoterInvitationData } from '../templates/voterInvitation';
import { PollResultsData } from '../templates/pollResults';

const router = Router();

// Validation schemas
const VoterInvitationSchema = z.object({
  pollTitle: z.string().min(1),
  pollDescription: z.string().optional(),
  votingUrl: z.string().url(),
  pollOwnerName: z.string().min(1),
  pollOwnerEmail: z.string().email(),
  closesAt: z.string().optional(),
  voterName: z.string().optional(),
  to: z.string().email()
});

const BulkVoterInvitationSchema = z.object({
  pollTitle: z.string().min(1),
  pollDescription: z.string().optional(),
  votingUrl: z.string().url(),
  pollOwnerName: z.string().min(1),
  pollOwnerEmail: z.string().email(),
  closesAt: z.string().optional(),
  recipients: z.array(z.object({
    email: z.string().email(),
    name: z.string().optional()
  })).min(1).max(100) // Limit bulk sends to 100 at a time
});

const PollResultsSchema = z.object({
  pollTitle: z.string().min(1),
  pollDescription: z.string().optional(),
  winnerName: z.string().min(1),
  totalVotes: z.number().min(0),
  resultsUrl: z.string().url(),
  pollOwnerName: z.string().min(1),
  voterName: z.string().optional(),
  finalRankings: z.array(z.object({
    position: z.number().min(1),
    name: z.string().min(1),
    votes: z.number().min(0),
    percentage: z.number().min(0).max(100)
  })),
  to: z.string().email()
});

// POST /api/email/voter-invitation - Send single voter invitation
router.post('/voter-invitation', async (req: Request, res: Response) => {
  try {
    const validated = VoterInvitationSchema.parse(req.body);
    
    const { to, ...invitationData } = validated;
    
    const result = await emailService.sendVoterInvitation(invitationData as VoterInvitationData, to);
    
    if (result.success) {
      res.json({
        success: true,
        data: {
          messageId: result.messageId,
          recipient: to
        }
      });
    } else {
      res.status(500).json({
        success: false,
        error: {
          code: 'EMAIL_SEND_FAILED',
          message: result.error || 'Failed to send email'
        }
      });
    }
  } catch (error) {
    if (error instanceof z.ZodError) {
      res.status(400).json({
        success: false,
        error: {
          code: 'VALIDATION_ERROR',
          message: 'Invalid request data',
          details: error.errors
        }
      });
    } else {
      console.error('Error sending voter invitation:', error);
      res.status(500).json({
        success: false,
        error: {
          code: 'INTERNAL_ERROR',
          message: 'Internal server error'
        }
      });
    }
  }
});

// POST /api/email/bulk-voter-invitations - Send bulk voter invitations
router.post('/bulk-voter-invitations', async (req: Request, res: Response) => {
  try {
    const validated = BulkVoterInvitationSchema.parse(req.body);
    
    const { recipients, ...baseData } = validated;
    
    // Send emails in batches to avoid overwhelming the email service
    const batchSize = 10;
    const batches = [];
    
    for (let i = 0; i < recipients.length; i += batchSize) {
      batches.push(recipients.slice(i, i + batchSize));
    }
    
    let totalSent = 0;
    let totalFailed = 0;
    const failedRecipients: string[] = [];
    
    for (const batch of batches) {
      const batchEmails = batch.map(r => r.email);
      const results = await emailService.sendBulkVoterInvitations(baseData as VoterInvitationData, batchEmails);
      
      totalSent += results.success.length;
      totalFailed += results.failed.length;
      
      // Track failed recipients
      results.failed.forEach((_, index) => {
        failedRecipients.push(batch[index].email);
      });
      
      // Small delay between batches to be respectful
      if (batches.length > 1) {
        await new Promise(resolve => setTimeout(resolve, 1000));
      }
    }
    
    res.json({
      success: true,
      data: {
        totalRecipients: recipients.length,
        sent: totalSent,
        failed: totalFailed,
        failedRecipients: failedRecipients.length > 0 ? failedRecipients : undefined
      }
    });
    
  } catch (error) {
    if (error instanceof z.ZodError) {
      res.status(400).json({
        success: false,
        error: {
          code: 'VALIDATION_ERROR',
          message: 'Invalid request data',
          details: error.errors
        }
      });
    } else {
      console.error('Error sending bulk voter invitations:', error);
      res.status(500).json({
        success: false,
        error: {
          code: 'INTERNAL_ERROR',
          message: 'Internal server error'
        }
      });
    }
  }
});

// POST /api/email/poll-results - Send poll results notification
router.post('/poll-results', async (req: Request, res: Response) => {
  try {
    const validated = PollResultsSchema.parse(req.body);
    
    const { to, ...resultsData } = validated;
    
    const result = await emailService.sendPollResults(resultsData as PollResultsData, to);
    
    if (result.success) {
      res.json({
        success: true,
        data: {
          messageId: result.messageId,
          recipient: to
        }
      });
    } else {
      res.status(500).json({
        success: false,
        error: {
          code: 'EMAIL_SEND_FAILED',
          message: result.error || 'Failed to send email'
        }
      });
    }
  } catch (error) {
    if (error instanceof z.ZodError) {
      res.status(400).json({
        success: false,
        error: {
          code: 'VALIDATION_ERROR',
          message: 'Invalid request data',
          details: error.errors
        }
      });
    } else {
      console.error('Error sending poll results:', error);
      res.status(500).json({
        success: false,
        error: {
          code: 'INTERNAL_ERROR',
          message: 'Internal server error'
        }
      });
    }
  }
});

// GET /api/email/status/:trackingId - Get email delivery status
router.get('/status/:trackingId', (req: Request, res: Response) => {
  const { trackingId } = req.params;
  
  const status = emailService.getEmailStatus(trackingId);
  
  if (status) {
    res.json({
      success: true,
      data: status
    });
  } else {
    res.status(404).json({
      success: false,
      error: {
        code: 'NOT_FOUND',
        message: 'Email tracking ID not found'
      }
    });
  }
});

// GET /api/email/tracking - Get all email tracking info (admin endpoint)
router.get('/tracking', (req: Request, res: Response) => {
  const allTracking = emailService.getAllEmailTracking();
  
  res.json({
    success: true,
    data: {
      total: allTracking.length,
      tracking: allTracking
    }
  });
});

// POST /api/email/retry-failed - Retry failed emails
router.post('/retry-failed', async (req: Request, res: Response) => {
  try {
    const result = await emailService.retryFailedEmails();
    
    res.json({
      success: true,
      data: result
    });
  } catch (error) {
    console.error('Error retrying failed emails:', error);
    res.status(500).json({
      success: false,
      error: {
        code: 'INTERNAL_ERROR',
        message: 'Failed to retry emails'
      }
    });
  }
});

// POST /api/email/cleanup - Clean up old tracking data
router.post('/cleanup', (req: Request, res: Response) => {
  try {
    emailService.cleanupTracking();
    
    res.json({
      success: true,
      data: {
        message: 'Tracking data cleaned up successfully'
      }
    });
  } catch (error) {
    console.error('Error cleaning up tracking data:', error);
    res.status(500).json({
      success: false,
      error: {
        code: 'INTERNAL_ERROR',
        message: 'Failed to cleanup tracking data'
      }
    });
  }
});

export { router as emailRouter };

File: backend/src/services/email.rs
use reqwest::Client;
use serde::{Deserialize, Serialize};
use anyhow::{Result, Context};

#[derive(Debug, Clone)]
pub struct EmailService {
    client: Client,
    base_url: String,
    api_key: String,
}

#[derive(Debug, Serialize)]
pub struct VoterInvitationRequest {
    #[serde(rename = "pollTitle")]
    pub poll_title: String,
    #[serde(rename = "pollDescription")]
    pub poll_description: Option<String>,
    #[serde(rename = "votingUrl")]
    pub voting_url: String,
    #[serde(rename = "pollOwnerName")]
    pub poll_owner_name: String,
    #[serde(rename = "pollOwnerEmail")]
    pub poll_owner_email: String,
    #[serde(rename = "closesAt")]
    pub closes_at: Option<String>,
    #[serde(rename = "voterName")]
    pub voter_name: Option<String>,
    pub to: String,
}

#[derive(Debug, Serialize)]
pub struct BulkVoterInvitationRequest {
    #[serde(rename = "pollTitle")]
    pub poll_title: String,
    #[serde(rename = "pollDescription")]
    pub poll_description: Option<String>,
    #[serde(rename = "votingUrl")]
    pub voting_url: String,
    #[serde(rename = "pollOwnerName")]
    pub poll_owner_name: String,
    #[serde(rename = "pollOwnerEmail")]
    pub poll_owner_email: String,
    #[serde(rename = "closesAt")]
    pub closes_at: Option<String>,
    pub recipients: Vec<EmailRecipient>,
}

#[derive(Debug, Serialize)]
pub struct EmailRecipient {
    pub email: String,
    pub name: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct PollResultsRequest {
    #[serde(rename = "pollTitle")]
    pub poll_title: String,
    #[serde(rename = "pollDescription")]
    pub poll_description: Option<String>,
    #[serde(rename = "winnerName")]
    pub winner_name: String,
    #[serde(rename = "totalVotes")]
    pub total_votes: usize,
    #[serde(rename = "resultsUrl")]
    pub results_url: String,
    #[serde(rename = "pollOwnerName")]
    pub poll_owner_name: String,
    #[serde(rename = "voterName")]
    pub voter_name: Option<String>,
    #[serde(rename = "finalRankings")]
    pub final_rankings: Vec<FinalRanking>,
    pub to: String,
}

#[derive(Debug, Serialize)]
pub struct FinalRanking {
    pub position: usize,
    pub name: String,
    pub votes: f64,
    pub percentage: f64,
}

#[derive(Debug, Deserialize)]
pub struct EmailResponse {
    pub success: bool,
    pub data: Option<EmailResponseData>,
    pub error: Option<EmailError>,
}

#[derive(Debug, Deserialize)]
pub struct EmailResponseData {
    #[serde(rename = "messageId")]
    pub message_id: Option<String>,
    pub recipient: Option<String>,
    pub sent: Option<usize>,
    pub failed: Option<usize>,
    #[serde(rename = "failedRecipients")]
    pub failed_recipients: Option<Vec<String>>,
}

#[derive(Debug, Deserialize)]
pub struct EmailError {
    pub code: String,
    pub message: String,
}

impl EmailService {
    pub fn new() -> Result<Self> {
        let base_url = std::env::var("EMAIL_SERVICE_URL")
            .unwrap_or_else(|_| "http://localhost:3001".to_string());
        
        let api_key = std::env::var("EMAIL_SERVICE_API_KEY")
            .context("EMAIL_SERVICE_API_KEY environment variable is required")?;

        Ok(Self {
            client: Client::new(),
            base_url,
            api_key,
        })
    }

    pub async fn send_voter_invitation(
        &self,
        request: VoterInvitationRequest,
    ) -> Result<EmailResponse> {
        let url = format!("{}/api/email/voter-invitation", self.base_url);
        
        let response = self
            .client
            .post(&url)
            .header("X-API-Key", &self.api_key)
            .json(&request)
            .send()
            .await
            .context("Failed to send HTTP request to email service")?;

        if !response.status().is_success() {
            let status = response.status();
            let text = response.text().await.unwrap_or_default();
            anyhow::bail!("Email service returned error {}: {}", status, text);
        }

        let email_response: EmailResponse = response
            .json()
            .await
            .context("Failed to parse email service response")?;

        Ok(email_response)
    }

    pub async fn send_bulk_voter_invitations(
        &self,
        request: BulkVoterInvitationRequest,
    ) -> Result<EmailResponse> {
        let url = format!("{}/api/email/bulk-voter-invitations", self.base_url);
        
        let response = self
            .client
            .post(&url)
            .header("X-API-Key", &self.api_key)
            .json(&request)
            .send()
            .await
            .context("Failed to send HTTP request to email service")?;

        if !response.status().is_success() {
            let status = response.status();
            let text = response.text().await.unwrap_or_default();
            anyhow::bail!("Email service returned error {}: {}", status, text);
        }

        let email_response: EmailResponse = response
            .json()
            .await
            .context("Failed to parse email service response")?;

        Ok(email_response)
    }

    pub async fn send_poll_results(
        &self,
        request: PollResultsRequest,
    ) -> Result<EmailResponse> {
        let url = format!("{}/api/email/poll-results", self.base_url);
        
        let response = self
            .client
            .post(&url)
            .header("X-API-Key", &self.api_key)
            .json(&request)
            .send()
            .await
            .context("Failed to send HTTP request to email service")?;

        if !response.status().is_success() {
            let status = response.status();
            let text = response.text().await.unwrap_or_default();
            anyhow::bail!("Email service returned error {}: {}", status, text);
        }

        let email_response: EmailResponse = response
            .json()
            .await
            .context("Failed to parse email service response")?;

        Ok(email_response)
    }

    pub async fn health_check(&self) -> Result<bool> {
        let url = format!("{}/health", self.base_url);
        
        let response = self
            .client
            .get(&url)
            .send()
            .await
            .context("Failed to send health check request")?;

        Ok(response.status().is_success())
    }
}

impl Default for EmailService {
    fn default() -> Self {
        Self::new().expect("Failed to create EmailService")
    }
}